File: android/src/main/java/org/ligi/scr/DefaultRetrofitCallback.java
package org.ligi.scr;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import retrofit2.Call;
import retrofit2.Callback;

abstract class DefaultRetrofitCallback<T> implements Callback<T> {

    private final boolean fatal;
    private final Activity activity;

    DefaultRetrofitCallback(boolean fatal, Activity activity) {
        this.fatal = fatal;
        this.activity = activity;
    }

    @Override
    public void onFailure(final Call<T> call, final Throwable t) {

        final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setTitle(android.R.string.dialog_alert_title);
        builder.setMessage(t.getMessage());

        if (fatal) {
            builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(final DialogInterface dialogInterface, final int i) {
                    dialogInterface.dismiss();
                    activity.finish();
                }
            });
        } else {
            builder.setPositiveButton(android.R.string.ok, null);
        }

        builder.show();
    }

}

File: android/src/main/java/org/ligi/scr/DefaultRetrofitCallback.kt
package org.ligi.scr

import android.app.Activity
import android.app.AlertDialog
import retrofit2.Call
import retrofit2.Callback

internal abstract class DefaultRetrofitCallback<T>(private val fatal: Boolean, private val activity: Activity) : Callback<T> {

    override fun onFailure(call: Call<T>, t: Throwable) {

        val builder = AlertDialog.Builder(activity)
        builder.setTitle(android.R.string.dialog_alert_title)
        builder.setMessage(t.message)

        if (fatal) {
            builder.setPositiveButton(android.R.string.ok) { dialogInterface, i ->
                dialogInterface.dismiss()
                activity.finish()
            }
        } else {
            builder.setPositiveButton(android.R.string.ok, null)
        }

        builder.show()
    }

}
